### DISCUSSION TITLE
Having debounce on sending data using webScoket

### DISCUSSION THREAD

**Discussion by majidsadr** [sourcelair/xterm.js#4524] [General]
I need to have something like debounce to send data to webSocket with a delay, but writing characters instantly to user. What I use now is a normal way to send data on entering each character:

```
    const { Terminal: XTerm } = await import("xterm");
    const xterm = new XTerm({
      cursorBlink: true,
      screenReaderMode: true,
      rows: 22,
      cols: 80,
      fontFamily: "Ubuntu Mono",
      fontSize: 18,
    });
    webSocketInstance.onopen = function (e) {
      xterm.onData((data) => {
        webSocketInstance.send("\x00" + data);
      });
   };
```

Is there any features or recommendations for this?
Thanks.

**Comment by jerch** [comment]
No thats not possible by normal means. The issue is the fact, that the character echoing is done by the PTY on server side, thus you need to send it over to the backend as fast as possible to get chars mirrored back in reasonable time.

What you are looking for is local echo mode, which xterm.js does not implement. So you'd have to implement that yourself, but thats very involved to get halfway right, as the program side might not simply echo input back, but trigger more complex display updates (example: hitting TAB at the shell prompt does not print a TAB, but gives you a list of executable commands). This cannot be resolved reliably, as you cannot know the action of a keypress in advance.

Edit: What some local echo handler do though - print what was locally typed, then wait for the PTY response and check if thats in line with what was printed. If not, erase local prints and display PTY response instead (I think mosh works similar to that).
